pkg/statistics/handle/usage/collector/collector.go

```diff
@@ -0,0 +1,164 @@
+}
+
+// NewGlobalCollector creates a new global collector
+func NewGlobalCollector[T any](mergeFn func(T), timeout time.Duration, channelSize int) GlobalCollector[T] {
+	g := &globalCollector[T]{
+		mergeFn:            mergeFn,
+		timeout:            timeout,
+		dataCh:             make(chan T, channelSize),
+		highPriorityDataCh: make(chan T, channelSize),
+		closeCh:            make(chan struct{}),
+	}
+	g.startWorker()
+	return g
+}
+
+// SpawnSession creates a related session collector from the global collector
+func (g *globalCollector[T]) SpawnSession() SessionCollector[T] {
+	return &sessionCollector[T]{
+		timeout:            g.timeout,
+		dataCh:             g.dataCh,
+		highPriorityDataCh: g.highPriorityDataCh,
+		lastUpdate:         time.Now(),
+	}
+}
+
+// startWorker spawns a goroutine to merge the data
+func (g *globalCollector[T]) startWorker() {
+	g.wg.Add(1)
+	go func() {
+		defer g.wg.Done()
+
+	loop:
+		for {
+			// nested selection to make sure `highPriorityDataCh` is selected before the normal `dataCh`
+			select {
+			case data := <-g.highPriorityDataCh:
+				g.mergeFn(data)
+			case <-g.closeCh:
+				break loop
+			default:
+				select {
+				case data := <-g.dataCh:
+					g.mergeFn(data)
+				case data := <-g.highPriorityDataCh:
+					g.mergeFn(data)
+				case <-g.closeCh:
+					break loop
+				}
+			}
+		}
+
+		// drain out the data from channel
+		g.flush()
+	}()
+}
+
+// flush reads all data from the channel, until the channel is empty
+func (g *globalCollector[T]) flush() {
+	for {
+		select {
+		case data := <-g.highPriorityDataCh:
+			g.mergeFn(data)
+		case data := <-g.dataCh:
+			g.mergeFn(data)
+		default:
+			return
+		}
+	}
+}
+
+// Close closes the background worker of the global collector
+func (g *globalCollector[T]) Close() {
+	g.closeOnce.Do(func() {
+		close(g.closeCh)
+		g.wg.Wait()
+	})
+}
+
+// SessionCollector is an interface to send stats data to the global collector
+type SessionCollector[T any] interface {
+	// SendDelta sends the data to the global collector. This function will not block (unless the `timeout` reached). It
+	// returns a bool to represent whether the data has been sent successfully.
+	SendDelta(data T) bool
+	// SendDeltaSync sends the data to the global collector. Unlike `SendDelta`, this function will always block and
+	// wait until the data has been received by the global collector.
+	SendDeltaSync(data T) bool
+}
+
+var _ SessionCollector[int] = &sessionCollector[int]{}
+
+// sessionCollector is the collector attached to each session to send the data to global collector
+type sessionCollector[T any] struct {
+	lastUpdate         time.Time
+	dataCh             chan<- T
+	highPriorityDataCh chan<- T
+	closeCh            <-chan struct{}
+	timeout            time.Duration
+}
+
+// SendDelta implements `SessionCollector[T]` interface
+func (s *sessionCollector[T]) SendDelta(data T) bool {
+	if time.Since(s.lastUpdate) > s.timeout {
+		return s.SendDeltaSync(data)
+	}
```

Comment:
Is this timeout necessary? Could we just remove it and make it simpler?

Reply:
After discussion, we agreed on hiding the argument `timeout`. (I also hide the argument `channelSize`).

I set a default value for them: timeout 5 minutes and channel size 10.